PrestaShop/PrestaShop
# Fix backoffice translations for third-party modules

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | Read below
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | maybe
| Deprecations? | no
| Fixed ticket? | maybe?
| How to test?  | Read below

# What this change is about

This PR basically fixes the translation BO for symfony modules that are NOT native. It also makes the new translation page available for any module, even if it's not symfony, with some caveats.

Normally, if a module is not in the default core catalogue (aka, if no file for it is found in `app/Resources/translations/default`), then it will fall back to the old 1.6.x translation page. Today, this is the case for all non-native modules.

The legacy translations page is only capable of "discovering" legacy translations, that is, calls to `l()` with no domain in `.php` and `.tpl` files. It works by parsing files using regexes, then saving the user's translations into `<locale>.php` files located within the module directory.

Since translations using the new translation system (`trans()`) cannot be discovered by this page, symfony-based modules were untranslatable -- until now.

To make your module translatable using the new system, you need to declare the following method on your main module class:

```php
public function isUsingNewTranslationSystem()
{
    return true;
}
```

Once that's done, when trying to translate your module in the BO, you'll be redirected to the new translation page, where the code will be analyzed for any new translations, and translations from any existing legacy translation file will be read as well.

Translations edited through this method are stored in the database.

You'll notice that domains are mysteriously "invented" by the backoffice, even if the code itself is still using calls to `l()` with no domain.

The domain convention is as follows:

- All modules belong to the `Modules` domain
- The specific domain for the module is constructed by taking the module name, removing any `ps_` prefix and any underscore. The first letter is in upper case, and the rest in lower case. This is handled by [`DomainHelper` in TranslationToolsBundle](https://github.com/PrestaShop/TranslationToolsBundle/blob/master/Translation/Helper/DomainHelper.php)
- The third part of the domain works the same as the legacy system: it is automatically decided according to the name of the file in which the wording was declared. Underscores are also removed, the first letter is in upper case, and the rest in lower case.

Of course, when translations are found using domains, the domain is left untouched.

# What this means

- Non-native symfony modules can now be translated through the BO
- Existing legacy module translations that aren't declared using domains will magically work in symfony module controllers as well, with no change in code needed.
- This means that a module could declare symfony controllers for 1.7.5+ and legacy controllers for previous versions, **all sharing the same translations**.

# What's missing

Translations are read from legacy `<locale>.php` files, but right now those files can only be created using the legacy page, which itself is incapable of discovering all wordings. This means that wordings translated using the new page are stored in database _only_, and therefore aren't exportable. Incidentally, this is also true for core wordings.

# What can be improved

- Saving the catalogue should write a translation file so that translations can be exported, or an export feature should be developed.
- Because of incompatibilities between the Symfony translator and the legacy translation files, the module's source code must be interpreted every time (and more than once per page load), which can be extremely slow. A cache system and cache invalidation via a button "Find wordings in code" (or something like that) would significantly speed things up.

# What's next

The legacy translator should be rewritten and tested.

# How to test it

1. Copy the module in `tests/Resources/modules/translation_test` to your `modules` directory
2. Go to your BO and install it (it's called "translation test")
3. Go to the Translations page in the BO, choose "installed modules translations" -> "translation test", pick your language
4. Some translations should be shown
5. Play around with the translations, they should be saved and loaded correctly
6. If you don't believe me, look in the `ps_translations` table


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/13504)
